internal/server/server.go

```diff
@@ -32,32 +31,62 @@ func Run(opts ...Option) error {
 		cfg.Logger = hclog.L()
 	}
 
-	// Setup our run group since we're going to be starting multiple
-	// goroutines for all the servers that we want to live/die as a group.
-	var group run.Group
-
-	// We first add an actor that just returns when the context ends. This
-	// will trigger the rest of the group to end since a group will not exit
-	// until any of its actors exit.
-	ctx, cancelCtx := context.WithCancel(cfg.Context)
-	cfg.Context = ctx
-	group.Add(func() error {
-		<-ctx.Done()
-		return ctx.Err()
-	}, func(error) { cancelCtx() })
+	log := cfg.Logger.Named("server")
 
-	// Setup our gRPC server.
-	if err := grpcInit(&group, &cfg); err != nil {
+	grpcServer, err := newGrpcServer(&cfg)
+	if err != nil {
 		return err
 	}
 
-	// Setup our HTTP server.
-	if err := httpInit(&group, &cfg); err != nil {
-		return err
+	wg := sync.WaitGroup{}
+
+	gprcErrs := make(chan error, 1)
+	wg.Add(1)
+	go func() {
+		err := grpcServer.start()
+		gprcErrs <- err
+		log.Debug("gRPC server has exited", "err", err)
+		wg.Done()
+	}()
+
+	httpErrs := make(chan error, 1)
+	httpEnabled := cfg.HTTPListener != nil
+	var httpServer *httpServer
+	if httpEnabled {
+		wg.Add(1)
+		httpServer = newHttpServer(grpcServer.server, &cfg)
+		go func() {
+			err := httpServer.start()
+			httpErrs <- err
+			log.Debug("HTTP server has exited", "err", err)
+			wg.Done()
+		}()
 	}
 
-	// Run!
-	return group.Run()
+	ctx, cancel := context.WithCancel(cfg.Context)
+	defer cancel()
+
+	defer wg.Wait() // Wait for server run processes to exit before returning
+	select {
+	case err := <-gprcErrs:
```

Comment:
This is all fine, another option I usually do (esp when there are a lot more goroutines to juggle) is to make `Close` idempotent for each one, then whenever there is an error, I don't care where it came from, I just call close again on all of them. If they're already closed its fine.